include/graphblas/reference/blas3.hpp

```diff
@@ -1111,92 +1118,545 @@ namespace grb {
+                                    assigns1 = 0;
+                                }
+                            }
+#else
+							(void)coors1.assign( k_col );
+#endif
+							(void)++nzc;
+						}
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+	                    while( !coors1.joinUpdate( local_update1 ) ) {}
+#endif
+
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+						#pragma omp barrier
+
+						#pragma omp for simd schedule( dynamic, config::CACHE_LINE_SIZE::value() )
+#endif
+						for( size_t l = B_raw.col_start[ i ]; l < B_raw.col_start[ i + 1 ]; ++l ) {
+							const size_t l_col = B_raw.row_index[ l ];
+							if( !coors1.assigned( l_col ) ) {
+								(void)++nzc;
+							}
+						}
 					}
+				}
+
+				const RC ret = grb::resize( C, nzc );
 #ifdef _DEBUG
-					std::cout << "are multiplied pairwise with ";
+				std::cout << "grb::resize( C, " << nzc << " ) = " << ret << "\n";
 #endif
+				return ret;
+			}
+
+			// computational phase
+			if( phase == EXECUTE ) {
+				// retrieve additional buffer
+				auto* const C_col_index = getReferenceBuffer< config::NonzeroIndexType >( n + 1 );
+
+				// perform column-wise nonzero count
+				nzc = 0;
+				for( size_t i = 0; i < m; ++i ) {
+					coors1.clear();
+					for( size_t k = A_raw.col_start[ i ]; k < A_raw.col_start[ i + 1 ]; ++k ) {
+						const size_t k_col = A_raw.row_index[ k ];
+						(void)coors1.assign( k_col );
+						(void)++nzc;
+
+						if( !crs_only ) {
+							(void) ++CCS_raw.col_start[ k_col + 1 ];
+						}
+					}
 					for( size_t l = B_raw.col_start[ i ]; l < B_raw.col_start[ i + 1 ]; ++l ) {
 						const size_t l_col = B_raw.row_index[ l ];
-						if( coors1.assigned( l_col ) ) {
-							coors2.assign( l_col );
-							(void)grb::apply( valbuf[ l_col ], valbuf[ l_col ], B_raw.getValue( l,
-								mulMonoid.template getIdentity< typename Operator::D2 >() ), oper );
+						if( !coors1.assigned( l_col ) ) {
+							(void)++nzc;
+							if( !crs_only ) {
+								(void)++CCS_raw.col_start[ l_col + 1 ];
+							}
+						}
+					}
+				}
+
+				// check capacity
+				if( nzc > capacity( C ) ) {
 #ifdef _DEBUG
-							std::cout << "B( " << i << ", " << l_col << " ) = " << B_raw.getValue( l,
-								mulMonoid.template getIdentity< typename Operator::D2 >() )
-							<< " to yield C( " << i << ", " << l_col << " ), ";
+					std::cout << "\t detected insufficient capacity "
+						<< "for requested operation\n";
+#endif
+					const RC clear_rc = clear( C );
+					if( clear_rc != SUCCESS ) {
+						return PANIC;
+					}
+					return FAILED;
+				}
+
+				// prefix sum for CCS_raw.col_start
+				if( !crs_only ) {
+					for( size_t j = 1; j < n; ++j ) {
+						CCS_raw.col_start[ j + 1 ] += CCS_raw.col_start[ j ];
+					}
+#ifndef NDEBUG
+					assert( CCS_raw.col_start[ 0 ] == 0 );
+					assert( CCS_raw.col_start[ n ] == nzc );
+#endif
+				}
+
+				// set C_col_index to all zero
+				if( !crs_only ) {
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+					#pragma omp parallel for simd
```

Comment:
explicitly select static schedule with right blocksize (use of OMP::localRange is preferred)

(simd can stay as per above also)